rse-r-package.Rmd

```diff
@@ -0,0 +1,590 @@
+
+The answer to the question in this section's title depends on what stage of development you are in.
+If you are doing [exploratory programming](glossary.html#exploratory-programming),
+a short comment to remind yourself of each function's purpose is good enough.
+(In fact, it's probably better than what most people do.)
+That comment should begin with an active verb and describe
+how inputs are turned into outputs.
+If the function has any [side effects](glossary.html#side-effects),
+you should eliminate them.
+If you can't,
+you should describe them too.
+
+An active verb is something like "extract", "normalize", or "find".
+For example,
+these are all good one-line comments:
+
+-   "Create a list of current ages from a list of birth dates."
+-   "Clip signals to lie in [0...1]."
+-   "Reduce the red component of each pixel."
+
+You can tell your one-liners are useful
+if you can read them aloud in the order the functions are called
+in place of the function's name and parameters.
+
+Once you start writing code for other people---including yourself three months from now---your
+documentation should describe:
+
+1.  The name and purpose of every function and constant in your code.
+2.  The name, purpose, and default value (if any) of every parameter to every function.
+3.  Any side effects the function has.
+4.  The type of value returned by every function.
+5.  Why and how the function will deliberately fail.
+    If a function uses something like `stopifnot` or `assert` to check that a condition holds,
+    then that halt-and-catch-fire behavior is effectively part of its interface.
+
+### Exercise: Fixing documentation
+
+FIXME: provide poorly-documented function and ask learner to find and fix gaps.
+
+### Exercise: Cross-references
+
+1.  What should you add to the roxygen2 comments for one function to link to the documentation for another function?
+2.  Add a cross-reference from the documentation for the function FIXME in the CO2 package to the documentation for the function FIXME.
+
+### Exercise: Documenting error conditions
+
+The guidelines above said that authors should document why and how their functions will deliberately fail.
+Where and how should you do this using roxygen2 for R?
+
+## How do I manage package dependencies? {#rse-r-package-dependencies}
+
+Let's add another function in another file.
+(We don't really need the function,
+and it doesn't have to go in another file,
+but we'll create it for tutorial purposes.)
+
+```{r, child="rse-r-package/etc/counter-original.R"}
+```
+
+This function depends on functions from the tidyverse---more specifically,
+from the dplyr package.
+Since our compiled-and-distributable package will only contain the compiled code for its own functions,
+direct calls to functions from other libraries won't work after the package is installed.
+To fix this,
+we add this to the roxygen2 comment block for our function:
+
+```{r eval=FALSE}
+#' @import dplyr
+#' @importFrom magrittr %>%
```

Comment:
Better yet, use:

```
usethis::use_pipe()
```

And this is done for you :) plus more.

Reply:
Made an exercise.